### PR TITLE
feat(evals): eval suite infrastructure — auto-generation, compilation, suite runner (Phase 2)

### LINKED ISSUE #152: Eval Infrastructure: EvalSuite, compile, run, API (Phase 2)
## Goal

Build the EvalSuite as a first-class entity that lets us go from SOP to demo-ready agent in under an hour. Auto-generate eval suites from SOPs, run them against compiled or hand-written agents, pinpoint failures, fix, re-run. Prove it works end-to-end with a second use case (not WISMO).

## Out of Scope (Phase 3+)

| Item | When | Why deferred |
|---|---|---|
| Replay tests (from pinpointed failures) | Phase 3 | Requires historical transcript browsing + break point UI |
| Fix logs (audit trail) | Phase 3 | Requires replay tests to link to |
| Simulation runner (multi-turn user simulator) | Phase 3 | Phase 2 scores existing sessions; simulation generates them |
| Live transcript eval (post-hoc scoring) | Phase 3 | Manual trigger, needs webhook/cron design |
| UI for eval suites | When API is proven stable | API-only for now, CLI/test-driven |
| Compiler quality improvements | After evals exist to measure against | Evals first, then optimize compiler |
| pass^k metric (multiple trials) | Phase 4 | Single run is sufficient for v1 |
| tool_sequence evaluator | Phase 4, if needed | tool_called is sufficient; Anthropic warns against sequence checking |
| Agent versioning (agent_versions table) | Phase 5 | compiled_instructions column on agents is sufficient for now |

## Data Flow

The end-to-end pipeline for a single test case within a suite run:

```
existing session (from production, test, or simulation)
  → session + session_messages (the transcript — role: user/assistant/tool rows)
  → runEvalSuite(suiteId, sessionId, test case assertions)
  → for each test case: resolveAssertions → executeAssertions against session_messages
  → eval_suite_runs + eval_runs + eval_run_scores
```

The suite evaluates existing sessions — it does not generate new ones. Agent execution happens elsewhere (production conversations, test runs, or the simulation runner in Phase 3). The evaluators only see `session_messages` rows. This is what makes the scorer source-agnostic — it works identically for synthetic sessions, live sessions, and replay sessions.

## Two Paths for Suite Creation

1. **`initSuiteFromSop(orgId, agentId, sopId)`** — SOP-based initialization. `sopId` is required. Derives test cases + assertions from SOP steps and agent guardrails. Supports re-initialization (deletes auto-generated, preserves manual).

2. **`createSuite(orgId, { agentId, name, sopId? })`** — Manual creation. Creates an empty suite. User populates test cases and assertions via CRUD endpoints. `sopId` is optional metadata (no derivation).

These are separate service functions and API endpoints — they don't share implementation.

### Suite initialization from SOP (initSuiteFromSop)

1. `initSuiteFromSop(orgId, agentId, sopId)` creates an `eval_suites` row and returns the suite with test cases (each carrying their own assertions). `sopId` is required.
2. For each SOP step with a `tool` reference and `required: true`, an `eval_configs` row of type `tool_called` is auto-created (idempotent by name) and linked as a **test case assertion** with `required: true` — assigned to test cases whose SOP path includes that step.
3. For each SOP step with a `tool` reference and `required: false`, an `eval_configs` row of type `tool_called` is auto-created and linked as a **test case assertion** with `required: false` — assigned only to test cases that exercise the optional branch.
4. For each SOP step with an instruction and `required: true` that has no tool reference, an `eval_configs` row of type `llm_judge` is auto-created (behavior-focused criterion derived from instruction text) and linked as a test case assertion. Criterion evaluates behavioral evidence, not explicit statements.
5. For each guardrail in the agent's knowledge base, an `eval_configs` row of type `llm_judge` is created (criterion from guardrail content) and linked as a test case assertion — assigned to **all** test cases (guardrails always apply).
6. Auto-generated test cases represent different paths through the SOP (minimum 3: happy path, edge case, guardrail). Each test case carries only the assertions relevant to that path. Happy path = all required steps. Edge case = same required steps, different category. Guardrail = required instruction steps + optional tool steps.
7. Each test case has `input` (JSONB), `mock_tool_responses` (JSONB), `category`, `source` (auto/manual), and its own `assertions[]` linking to eval_configs.
8. Re-initializing a suite for the same agent+SOP deletes old auto-generated test cases (cascade deletes their assertions), creates new ones. Manually-added test cases (and their assertions) are preserved.

### Manual suite creation (createSuite)

9. `POST /api/eval-suites` with `{ agentId, name, sopId? }` creates an empty suite. No test cases or assertions are auto-generated.
10. User adds test cases via `POST /api/eval-suites/:suiteId/test-cases` and assertions via `POST /api/eval-suites/:suiteId/test-cases/:testCaseId/assertions`.

### Agent compilation (compileAgent)

11. `POST /api/compiler/agents/:agentId/compile` with `{ sopId }` compiles the SOP + guardrails into a system prompt via the existing compiler pipeline.
12. The compiled prompt is persisted to `agents.compiled_instructions`, with `compiled_at` timestamp and `compiled_from` JSONB metadata (`{ sopId, guardrailIds }`).
13. Response returns `{ compiledPrompt, compiledFrom, compiledAt }`.
14. If the agent or SOP doesn't exist, returns 404.
15. `agents.compiled_instructions` is the single instructions field — set by compilation via this endpoint or by direct `PATCH /api/agents/:id` update for hand-written prompts.

### Synthetic session storage (storeSyntheticSession)

16. `storeSyntheticSession({ orgId, agentId, generationResult, userInput, channelType })` is a production service that takes a Mastra `agent.generate()` result and persists it as a completed session.
17. The user input is stored as a `role: "user"` message via `addMessage()`.
18. For each step in the Mastra generation result: assistant text is stored as `role: "assistant"` message; each tool call is stored as a `role: "tool"` message with `tool_name`, `tool_input`, `tool_output`, and `tool_status: "success"`.
19. The session is marked `status: "completed"` after all messages are stored.
20. This replaces the test-only `storeSession()` helper — that helper delegates to this service.

### Run eval suite (runEvalSuite)

21. `POST /api/eval-suites/:suiteId/run` with body `{ sessionId, promptSource: "compiled" | "hand_written" | "edited" }` creates an `eval_suite_runs` row and evaluates the given session against each test case's assertions.
22. If the agent's `compiled_instructions` is null, returns 422.
23. For each test case: resolves assertions (loads eval_configs, resolves tool names at scoring time), then scores the session's `session_messages` against those assertions using the shared `executeAssertions` scorer.
24. Each `eval_runs` row created by a suite run has `sourceType: "suite"`, `sourceId: suite.id`, and `suite_run_id` FK pointing back to the `eval_suite_runs` row.
25. Suite run response includes per-test-case pass/fail and overall summary, aggregated from `eval_runs` at query time (no cached counters on `eval_suite_runs`).
26. If the suite has no test cases, returns 422. If any test case has no assertions, returns 422.
27. If the suite is archived, returns 409.

### Shared scorer

28. `executeAssertions` is the shared scoring function — accepts `ResolvedAssertion[]` and `session_messages`, produces `eval_run_scores` rows.
29. Tool name resolution (connectorToolId → runtime name) happens at scoring time inside `resolveAssertions`, not at suite init time.
30. The legacy `runEvaluation(orgId, sessionId, "sop", sopId)` path has been removed. Suite is the only eval entry point. `sourceType` enum is `["suite", "replay_test", "live"]`.

### EvalSuite CRUD

31. `POST /api/eval-suites/init` creates a suite from SOP via `initSuiteFromSop` — body: `{ agentId, sopId }`.
32. `POST /api/eval-suites` creates an empty suite via `createSuite` — body: `{ agentId, name, sopId? }`.
33. `GET /api/eval-suites/:suiteId` returns the suite with its test cases, each including assertions.
34. `GET /api/eval-suites/:suiteId/runs` returns suite runs with per-test-case results aggregated from `eval_runs`.
35. `GET /api/eval-suites/:suiteId/runs/:runId` returns a single run with detailed per-test-case scores.
36. `POST /api/eval-suites/:suiteId/test-cases` adds a manual test case with `source: "manual"`.
37. `POST /api/eval-suites/:suiteId/test-cases/:testCaseId/assertions` adds a manual assertion with `source: "manual"`.
38. `DELETE /api/eval-suites/:suiteId` deletes a suite and all children (cascade).
39. Suite not found returns 404. Validation errors return 422.

### Schema changes

40. New `eval_suites` table with RLS enabled, org-scoped: `id`, `organization_id`, `agent_id` (FK), `sop_id` (FK, nullable), `name`, `status` (active/archived), `description`, `created_by`, `created_at`, `updated_at`.
41. New `eval_suite_test_cases` table with FK to `eval_suites` (cascade delete), org-scoped + RLS: `id`, `organization_id`, `suite_id`, `name`, `category`, `description`, `source` (auto/manual), `input` (JSONB), `expected_behavior`, `mock_tool_responses` (JSONB), `order`, `created_at`, `updated_at`.
42. New `eval_suite_assertions` table with FK to `eval_suite_test_cases` (cascade) and FK to `eval_configs` (no cascade), org-scoped + RLS: `id`, `organization_id`, `test_case_id`, `eval_config_id`, `name`, `sop_step_id` (nullable), `source` (auto/manual), `order`, `required`, `created_at`.
43. New `eval_suite_runs` table (thin aggregator, no cached counters), org-scoped + RLS: `id`, `organization_id`, `suite_id`, `prompt_source`, `triggered_by`, `started_at`, `completed_at`, `duration_ms`.
44. `eval_runs` table gets nullable `suite_run_id` FK to `eval_suite_runs` + nullable `test_case_id`.
45. `agents` table gets `compiled_instructions TEXT`, `compiled_at TIMESTAMPTZ`, `compiled_from JSONB`.
46. All new tables have RLS enabled and `organization_id` FK.
47. `eval_source_type` enum extended: `["suite", "replay_test", "live"]` (legacy `"sop"` value remains in DB but is unused by code).

### Regression + dry run

48. WISMO E2E test: `initSuiteFromSop` → `compileAgent` → run agent → `storeSyntheticSession` → `runEvalSuite` → assertions pass (tool_called, llm_judge with behavioral criteria).
49. Second SOP (appointment booking) dry run: validates auto-generation is SOP-agnostic.

### Error handling

50. `initSuiteFromSop` for a non-existent agent or SOP returns 404.
51. `runEvalSuite` on an archived suite returns 409.
52. `runEvalSuite` when test case has no assertions returns 422.
53. `runEvalSuite` when agent has no `compiled_instructions` returns 422.

## Constraints

- Must reuse existing evaluator implementations (`tool_called`, `no_tool_called`, `tool_input_contains`, `llm_judge`) — no new evaluator types.
- Must reuse existing `eval_runs` + `eval_run_scores` tables for per-test-case results (no parallel scoring tables).
- Compiled output stored on `agents` table for now — migrates to `agent_versions` in Phase 5.
- `agents.compiled_instructions` is the single instructions field for both compiled and hand-written prompts.
- LLM judge model configured via existing `EVAL_LLM_MODEL` env var (default Claude Haiku).
- Mastra is already a dependency (PR #151) — acceptable for production use.
- All new tables must have RLS enabled with `organization_id` FK.
- Suite auto-generation must be deterministic (same SOP → same assertions + test case structure).
- Tool name resolution (connectorToolId → runtime name) must happen at scoring time, not at suite init time — prevents stale mappings.
- Suite evaluates existing sessions — it does not generate transcripts (simulation runner is Phase 3).

## Dependencies

- `@features/compiler/` — PR #151, merged. Provides `compile()` and `toMastra()`.
- `@features/evals/` — Production. Provides evaluator registry, `executeAssertions()`, `eval_configs` CRUD, `eval_runs` + `eval_run_scores` schema.
- `@features/sops/` — Production. SOP CRUD, `sop_steps` with `SopStep` type.
- `@features/knowledge-base/` — PR #140, merged. Provides guardrails for assertion derivation.
- `@features/sessions/` — Production. `createSession()`, `addMessage()`, `updateSession()`.
- Mastra SDK (`@mastra/core`) — installed via PR #151. Used for `agent.generate()` in E2E tests.

## Technical Approach

**Data model** — Four new tables + two ALTER TABLEs. `eval_suites` is the parent (`sopId` nullable for manual suites); `eval_suite_test_cases` cascades from it; `eval_suite_assertions` cascades from `eval_suite_test_cases` (assertions belong to test cases, not suites — each test case carries only the assertions relevant to the SOP path it exercises). `eval_suite_assertions` has a FK to `eval_configs` (no cascade, since configs may be shared) and a `required` boolean column. `eval_suite_runs` is a thin aggregator with no cached counters. `eval_runs` gets `suite_run_id` + `test_case_id` FKs. `agents` gets three compiled_* columns.

**Two creation paths** — `initSuiteFromSop` (SOP required, derives everything) and `createSuite` (manual, empty suite). These are separate functions with separate routes — they don't share implementation because they solve different problems.

**Shared scorer** — `executeAssertions()` in `evals.service.ts` is the canonical scoring function. It accepts `ResolvedAssertion[]` and `session_messages`, runs each assertion through the evaluator registry, produces `eval_run_scores` rows. The legacy `runEvaluation()` SOP-based path has been removed — suite is the only eval entry point.

**initSuiteFromSop service** — Loads SOP steps and guardrails. For each step, auto-creates eval_configs (idempotent by name). Generates 3 path-based test cases (happy/edge/guardrail) and assigns step assertions to each based on which SOP path the test case exercises. Guardrail assertions go on all test cases. `llm_judge` criteria are behavior-focused ("demonstrates correct performance") rather than literal ("followed this instruction").

**storeSyntheticSession service** — Production service that bridges agent runner and eval scorer. Takes Mastra `agent.generate()` result, creates a session, stores messages (user input + assistant text + tool calls), marks completed.

**compileAgent service** — Thin wrapper around `compile()` + `toMastra()`. Loads SOP + guardrails, runs compiler pipeline, persists to `agents.compiled_instructions`.

**runEvalSuite service** — Takes `suiteId` + `sessionId`. Validates suite state (not archived, has test cases with assertions, agent has instructions). For each test case: `resolveAssertions()` (loads eval_configs, resolves tool names at scoring time) → `executeAssertions()` against session messages → persists eval_run + eval_run_scores. Aggregates at query time.

### Why assertions belong to test cases, not suites?

Different test cases exercise different SOP paths. A happy-path test case should check `tool_called(store_look_up_order)` but not `tool_called(helpdesk_create_ticket)`. A guardrail test case should check the opposite. Suite-level assertions force every test case through the same checks, which is wrong for branching SOPs.

### Why two separate creation paths?

SOP-based suites and manual suites are different animals. SOP suites derive everything from the spec — auto-generate test cases, auto-create eval_configs, assign assertions by path. Manual suites are empty shells populated by the user. Mixing these in one function creates confusion about what's required and what's derived.

### Why behavior-focused llm_judge criteria?

The auto-generated criterion `"The agent followed this instruction: \"Classify the email intent\""` is too literal — the LLM judge expects explicit classification when the agent classifies implicitly by choosing the right path. Behavior-focused criteria (`"The agent's response demonstrates correct performance of this step"`) evaluate what the agent did, not what it said about doing it.

### Why suite scores existing sessions, not generating new ones?

The suite is a scorer, not a simulator. Session generation (building Mastra agents, running `agent.generate()`, creating mock tools) is the simulation runner's job (Phase 3). Keeping them separate means the suite can score any session — production, test, simulated, replayed — without caring how it was created.

### Why remove the legacy "sop" eval path?

Once suites exist, scoring against raw SOP steps is redundant. The suite IS the SOP-derived eval, but better — it supports manual assertions, path-based test cases, and re-initialization. One eval path is simpler than two.

### Why eval_suite_runs is a thin aggregator?

Cached counters rot. For 3-15 test cases per suite, query-time aggregation from `eval_runs WHERE suite_run_id = ?` is negligible. This keeps `eval_runs` + `eval_run_scores` as the single source of truth.

### Why resolve tool names at scoring time?

Storing tool name mappings at suite init time creates stale data if connectors are reconfigured. Resolving at scoring time ensures evaluators always use current connector_tools + connectors configuration.

## Sequencing Roadmap

| Step | What | Dependency | Scope |
|---|---|---|---|
| 1 | SOP-to-Agent Compiler Phase 1 | PR #151 (merged) | Done |
| 2 | Eval Infrastructure (suites, scorer, compile, API) | Step 1 | **This PRD** |
| 3 | Improvement Machine (replay tests, fix logs, simulation runner) | Step 2 | Phase 3 |
| 4 | Harden + Extend (pass^k, tool_sequence, 15 test cases, Langfuse) | Step 3 | Phase 4 |
| 5 | Agent Versioning (agent_versions table) | Step 2 | Phase 5 |

### PR DESCRIPTION
**Pull request by pfijalki**
## Summary

Adds EvalSuite as a first-class entity for automated agent quality evaluation. Two paths: SOP-based initialization (auto-generates test cases + assertions from SOP steps and guardrails) and manual suite creation (empty suite, user populates via CRUD).

## Changes

- **Schema**: 4 new tables (`eval_suites`, `eval_suite_test_cases`, `eval_suite_assertions`, `eval_suite_runs`) + `compiled_*` columns on `agents` + `suite_run_id`/`test_case_id` on `eval_runs`
- **Compiler service**: `POST /api/compiler/agents/:agentId/compile` — SOP + guardrails → system prompt, persists to `agents.compiled_instructions`
- **Eval suite service**: `initSuiteFromSop` (SOP-based) + `createSuite` (manual) + `runEvalSuite` (scores existing sessions) + CRUD for test cases/assertions
- **Synthetic session service**: `storeSyntheticSession` — Mastra agent result → session_messages (production service, replaces test helper)
- **Shared scorer**: `executeAssertions` reused by both suite and legacy paths; tool name resolution at scoring time
- **Routes**: Full CRUD + run + runs listing under `/api/eval-suites`
- **Removed legacy `"sop"` eval path** — suite is the only eval entry point; `sourceType` enum now `["suite", "replay_test", "live"]`

## Key Design Decisions

- **Assertions belong to test cases, not suites** — each test case carries only the assertions relevant to its SOP path (happy path vs escalation)
- **Suite scores existing sessions** — does not generate transcripts (simulation runner is Phase 3)
- **`sopId` optional on suites** — supports manual suites without SOP derivation
- **Behavior-focused llm_judge criteria** — evaluates agent actions, not explicit statements
- **No cached counters** — `eval_suite_runs` aggregates pass/fail at query time from `eval_runs`

## Test Coverage

- E2E: WISMO suite — `initSuiteFromSop` → `compileAgent` → run agent → `storeSyntheticSession` → `runEvalSuite` → assertions pass
- E2E: Appointment booking SOP — validates auto-generation is SOP-agnostic
- Integration: Error handling (404, 409, 422 cases)
- Unit: Typecheck + lint clean

Closes #152

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: modelguide-api/drizzle/0028_eval-suites.sql [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+CREATE POLICY "eval_suite_runs_org_isolation" ON "eval_suite_runs"
+  USING ("organization_id" = current_setting('app.organization_id')::uuid);
+CREATE POLICY "eval_suite_runs_org_insert" ON "eval_suite_runs"
+  FOR INSERT WITH CHECK ("organization_id" = current_setting('app.organization_id')::uuid);
+
+-- ============================================================================
+-- Extend eval_runs with suite linkage columns
+-- ============================================================================
+
+ALTER TABLE "eval_runs" ADD COLUMN IF NOT EXISTS "suite_run_id" uuid;
```
Comment (pekunicki) [inline modelguide-api/drizzle/0028_eval-suites.sql]: `suite_run_id` and `test_case_id` are added as plain UUID columns, but the service and ADR treat them as real relationships. Without foreign keys we can create dangling eval runs and stale suite-run links. Can we add references to `eval_suite_runs(id)` and `eval_suite_test_cases(id)` here?

File: docs/decisions/009-eval-suites.md [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+├── TestCase: "Edge case" (category: edge_case)
+│   ├── assertion: step:1:llm_judge (classify) — required: true
+│   ├── assertion: step:2:llm_judge (extract)  — required: true
+│   ├── assertion: step:3:tool_called (store_look_up_order) — required: true
+│   └── assertion: step:4:llm_judge (compose)  — required: true
+│
+└── TestCase: "Guardrail path" (category: guardrail)
+    ├── assertion: step:1:llm_judge (classify) — required: true
+    ├── assertion: step:2:llm_judge (extract)  — required: true
+    └── assertion: step:5:tool_called (helpdesk_create_ticket) — required: false
```
Comment (pekunicki) [inline docs/decisions/009-eval-suites.md]: not sure if i'd call it guardrail path, maybe unhappy scenario, since guardrails is something different - i.e. never call customer by surname etc.

File: docs/decisions/009-eval-suites.md [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+│   ├── Load session_messages for sessionId
+│   ├── executeAssertions(resolved, messages) → eval_run_scores
+│   └── Update eval_runs (passed, durationMs, completedAt)
+│
+└── Update eval_suite_runs (completedAt, durationMs)
+```
+
+### Legacy SOP eval path removed
+
+The `runEvaluation(orgId, sessionId, "sop", sopId)` function and `POST /api/evals/runs` endpoint have been removed. Suite is the single eval entry point. The `eval_source_type` enum is now `["suite", "replay_test", "live"]` — `"sop"` remains in the DB enum but is unused by code.
```
[the remark this answers is not in the document]
Comment (pekunicki) [reply docs/decisions/009-eval-suites.md]: awesome, I'm slightly towards option 1, so SOPs don't become too complex
